src/components/application_manager/include/application_manager/app_service_manager.h

```diff
@@ -0,0 +1,124 @@
+ * may be used to endorse or promote products derived from this software
+ * without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef SRC_COMPONENTS_APPLICATION_MANAGER_INCLUDE_APPLICATION_MANAGER_APP_SERVICE_MANAGER_H_
+#define SRC_COMPONENTS_APPLICATION_MANAGER_INCLUDE_APPLICATION_MANAGER_APP_SERVICE_MANAGER_H_
+
+#include "smart_objects/smart_object.h"
+
+namespace application_manager {
+
+struct AppService {
+  uint32_t connection_key;
+  bool mobile_service;
+  bool default_service = false;
+  smart_objects::SmartObject record;
+};
+
+class ApplicationManager;
+
+/**
+ * @brief The AppServiceManager is TODO.
+ */
+class AppServiceManager {
+ public:
+  /**
+   * @brief Class constructor
+   * @param app_manager
+   */
+  AppServiceManager(ApplicationManager& app_manager);
+
+  /**
+   * @brief Class destructor
+   */
+  ~AppServiceManager();
+
+  /**
+   * @brief TODO
+   * @param manifest
+   */
+  smart_objects::SmartObject PublishAppService(
+      const smart_objects::SmartObject& manifest,
+      const bool mobile_service,
+      const uint32_t connection_key);
+
+  /**
+   * @brief TODO
+   * @param service_id
+   */
+  bool UnpublishAppService(const std::string service_id);
+
+  /**
+   * @brief TODO
+   * @param service_id
+   */
+  bool SetDefaultService(const std::string service_id);
+
+  /**
+   * @brief TODO
+   * @param service_id
+   */
+  bool RemoveDefaultService(const std::string service_id);
+
+  /**
+   * @brief TODO
+   * @param service_id
+   */
+  bool ActivateAppService(const std::string service_id);
+
+  /**
+   * @brief TODO
+   * @param service_id
+   */
+  bool DeactivateAppService(const std::string service_id);
+
+  /**
+   * @brief TODO
+   * @param manifest
+   */
+  std::vector<smart_objects::SmartObject> GetAllServices();
+
+  std::pair<std::string, AppService> ActiveServiceByType(
+      std::string service_type);
+
+  std::pair<std::string, AppService> FindServiceByName(std::string name);
+
+  std::string DefaultServiceByType(std::string service_type);
+
+ private:
+  ApplicationManager& app_manager_;
+  std::map<std::string, AppService> published_services_;
```

Comment:
This really needs a lock, could be implemented once all of the pieces of the feature are merged into this branch

Reply:
Fixed with https://github.com/smartdevicelink/sdl_core/pull/2802/commits/090ebab24ddb3c761386f20e873b0142fb42db51